### PR TITLE
Use Theano's CuDNN implementation of batch normalization

### PR DESCRIPTION
**Pull request by crowsonkb**
Following on from https://github.com/fchollet/keras/pull/3260, use Theano's new CuDNN implementation of batch normalization, falling back if it is unavailable or we are not running on a GPU.

### CODE REVIEW

**Comment by fchollet** [comment]
This should be refactored as:

```python
try
   return 
except
   pass
return
```

And it should be tested.

**Comment by crowsonkb** [comment]
What tests exactly do I need?

**Comment by fchollet** [comment]
I mean, did you test it, i.e. did you check that the output is the same with both implementations?

Since we cannot run cuDNN tests on Travis there is no need to add a new unit test.

**Comment by crowsonkb** [comment]
So far I've checked that I'm able to train models as expected with both implementations and that the existing unit tests pass, not that the output is close or the same.

**Comment by fchollet** [comment]
I was specifically concerned about whether the `var` argument matches the previous implementation or not. This is not something you cannot verify by training a new network from scratch. You could try training a network with the old implementation, then saving it to disk and verifying that the outputs are the same at test time after loading it and running it with cuDNN.

**Comment by crowsonkb** [comment]
```python
from keras.layers import BatchNormalization, Input
from keras.models import Model
import numpy as np

np.random.seed(0)
bn_input = Input(shape=(4, 4))
bn_output = BatchNormalization(axis=1)(bn_input)
bn = Model(bn_input, bn_output)
bn.compile('sgd', loss='mse')

data = np.random.normal(loc=10, scale=5, size=(4, 4, 4))
bn.fit(data, data)
out = bn.predict(data)
print(out.mean(), out.std())
```

This code produced similar results on GPU using CuDNN and on CPU: `6.34687 3.38879` on GPU and `6.34685 3.38877` with `THEANO_FLAGS=device=cpu`.

**Comment by fchollet** [comment]
You would observe this whether or not any changes were introduced, because these two quantities only depend on the the `var` and `mean` outputs of `normalize_batch_in_training`.

Try saving the model after training on GPU, loading it on CPU, and check that the predictions are the same.

**Comment by fchollet** [comment]
I misread it, sorry. Thought you were referring the `mean` and `std` attributes of the layer. Looks like it's working then.

**Comment by fchollet** [comment]
Actually, not necessarily. Do try the saving/loading test.

**Comment by crowsonkb** [comment]
After saving it on the GPU and loading it on the CPU:

```
In [6]: bn = keras.models.load_model('bn_test.h5')

In [7]: out = bn.predict(data)

In [8]: out.mean()
Out[8]: 6.3468485

In [9]: out.std()
Out[9]: 3.3887751
```

I'm not sure what the small differences are due to.

**Comment by fchollet** [comment]
I assume numeric accumulation is handled differently by cuDNN. The differences look normal to me.

LGTM, thanks for checking.

**Comment by mbotumbo** [comment]
@crowsonkb the catch is failing for me when cuDNN isn't detected. I get a untyped exception, not an AttributeError:

> Exception: ("We can't determine the cudnn version as it is not available", 'Can not compile with cuDNN. We got this error:\n/tmp/try_flags_h_BsVg.c:5:19: fatal error: cudnn.h: No such file or directory\n #include <cudnn.h>\n                   ^\ncompilation terminated.\n')

**Comment by crowsonkb** [comment]
@mbotumbo: I just tried temporarily moving my `cudnn.h`, making CuDNN unavailable, but couldn't reproduce this. Which platform/CUDA version/etc. are you using?

**Comment by mbotumbo** [comment]
@crowsonkb Sorry about that, I obviously should have included those details but was in a rush to get to work. 

ubuntu 14.04, python 2.7, CUDA release 7.5, V7.5.17, theano/kears updated before i posted.... anything else i'm forgetting?

**Comment by Dapid** [comment]
This seems to fail on FCN with a ValueError.

See here: https://groups.google.com/forum/#!topic/keras-users/innS4NcLPt8
